ClassFieldTheory/IsNonarchimedeanLocalField/Unramified.lean

```diff
@@ -0,0 +1,320 @@
+import ClassFieldTheory.Mathlib.FieldTheory.Finite.IntermediateField
+import ClassFieldTheory.Mathlib.RingTheory.HenselPolynomial
+import ClassFieldTheory.Mathlib.RingTheory.RootsOfUnity.PrimitiveRoots
+
+/-! # Unramified extension of local field of a given degree
+
+This file shows that if `K` is a non-archimedean local field and `n > 0` is any positive integer,
+then there is a unique (up to in general non-unique isomorphism) unramified extension of `K` of
+degree `n`.
+-/
+
+noncomputable section
+
+namespace IsNonarchimedeanLocalField
+
+open Polynomial ValuativeRel
+
+/-- **The** unramified extension of degree `n > 0` of a given non-archimedean local field `K`. -/
+def UnramifiedExtension (K : Type*) [Field K] [ValuativeRel K] [TopologicalSpace K]
+    [IsNonarchimedeanLocalField K] (n : ℕ) : Type _ :=
+  SplittingField (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1 : K[X])
+deriving Field, Algebra K, FiniteDimensional K
+
+variable (K : Type*) [Field K] [ValuativeRel K] [TopologicalSpace K] [IsNonarchimedeanLocalField K]
+variable (n : ℕ)
+
+instance : ValuativeRel (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose
+
+instance : ValuativeExtension K (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose_spec.choose
+
+instance : TopologicalSpace (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose_spec.choose_spec.choose
+
+instance : IsNonarchimedeanLocalField (UnramifiedExtension K n) :=
+  (isNonarchimedeanLocalField_of_finiteDimensional K _).choose_spec.choose_spec.choose_spec
+
+instance UnramifiedExtension.isSplittingField :
+    IsSplittingField K (UnramifiedExtension K n) (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1) :=
+  .splittingField _
+
+theorem UnramifiedExtension.splits :
+    (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1 : (UnramifiedExtension K n)[X]).Splits (.id _) := by
+  have := (UnramifiedExtension.isSplittingField K n).splits
+  simpa using (splits_id_iff_splits _).mpr this
+
+theorem UnramifiedExtension.factors :
+    (X ^ (Nat.card 𝓀[K] ^ n - 1) - 1 : (UnramifiedExtension K n)[X]).Factors := by
+  simpa [Splits] using splits K n
+
+open UnramifiedExtension
+
+section zero
+
+theorem finrank_unramifiedExtension_zero : Module.finrank K (UnramifiedExtension K 0) = 1 := by
+  have := UnramifiedExtension.isSplittingField K 0
+  rw [pow_zero, Nat.sub_self, pow_zero, sub_self, isSplittingField_iff_intermediateField,
+    rootSet_zero, IntermediateField.adjoin_empty,
+    IntermediateField.bot_eq_top_iff_finrank_eq_one] at this
+  exact this.2
+
+instance : Subsingleton (Subalgebra K (UnramifiedExtension K 0)) :=
+  subsingleton_of_bot_eq_top <| Subalgebra.bot_eq_top_of_finrank_eq_one <|
+    finrank_unramifiedExtension_zero K
+
+instance : Subsingleton (IntermediateField K (UnramifiedExtension K 0)) :=
+  IntermediateField.toSubalgebra_injective.subsingleton
+
+end zero
+
+-- An auxiliary lemma that we might need more than once
```

Comment:
You *do* need it more than once (you need it twice)

Reply:
done